src/tdycharacteristiccurves.c

```diff
@@ -231,6 +274,149 @@ void RelativePermeability_Mualem(PetscReal m,PetscReal Se,PetscReal *Kr,
   }
 }
 
+/* -------------------------------------------------------------------------- */
+///
+/// Sets up a cubic polynomial interpolation for relative permeability following
+/// PFLOTRAN's approach of smoothing relative permeability functions
+///
+/// @param [in] x1           low value of x
+/// @param [in] x2           high value of x
+/// @param [inout] *rhs      rhs vector (input) and coefficients (output)
+///
+///  f(x) = a0 + a1 * x + a2 * x^2 + a2 * x^3
+///  df_dx = a1 + 2 * a2 * x + 3 * a3 * x^2
+///
+/// Constraints:
+/// f(low)      = rel_perm_fn(x)      = f1
+/// f(high)     = 1.0                 = f2
+/// df_dx(low)  = drv_rel_perm_fn(x)  = df1_dx
+/// df_dy(high) = 0.0                 = df2_dx
+///
+/// Linear system:
+/// a0 + a1 * x1 + a2 * x1^2   + a2 * x1^3     = f1
+/// a0 + a1 * x2 + a2 * x1^2   + a2 * x2^3     = f2
+///      a1      + 2 * a2 * x1 + 3 * a3 * x1^2 = df1_dx
+///      a1      + 2 * a2 * x2 + 3 * a3 * x2^2 = df2_dx
+///
+PetscErrorCode CubicPolynomialSetup(PetscReal x1, PetscReal x2, PetscReal *rhs) {
```

Comment:
If you know that `rhs` should be an array of length 4, you can enforce that in the function declaration with `PetscReal rhs[4]`. That's also a nice cue to users.